azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -494,58 +494,62 @@ def _run_op(self, op):
 
             request_id = str(uuid.uuid4())
 
-            # Alias to avoid overload within the callback below
-            # CT-TODO: remove the need for this with better callback semantics
-            op_waiting_for_response = op
-
-            @pipeline_thread.runs_on_pipeline_thread
-            def on_send_request_done(op, error):
-                logger.debug(
-                    "{}({}): Finished sending {} request to {} resource {}".format(
-                        self.name,
-                        op_waiting_for_response.name,
-                        op_waiting_for_response.request_type,
-                        op_waiting_for_response.method,
-                        op_waiting_for_response.resource_location,
-                    )
-                )
-                if error:
-                    logger.debug(
-                        "{}({}): removing request {} from pending list".format(
-                            self.name, op_waiting_for_response.name, request_id
-                        )
-                    )
-                    del (self.pending_responses[request_id])
-                    op_waiting_for_response.complete(error=error)
-                else:
-                    # request sent.  Nothing to do except wait for the response
-                    pass
-
-            logger.debug(
-                "{}({}): Sending {} request to {} resource {}".format(
-                    self.name, op.name, op.request_type, op.method, op.resource_location
-                )
-            )
-
             logger.debug(
                 "{}({}): adding request {} to pending list".format(self.name, op.name, request_id)
             )
             self.pending_responses[request_id] = op
 
-            new_op = pipeline_ops_base.RequestOperation(
-                method=op.method,
-                resource_location=op.resource_location,
-                request_body=op.request_body,
-                request_id=request_id,
-                request_type=op.request_type,
-                callback=on_send_request_done,
-                query_params=op.query_params,
-            )
-            self.send_op_down(new_op)
+            self._send_request_down(request_id, op)
```

Comment:
I don't quite understand why this was moved up, if you could explain that to me I'd appreciate it.

Reply:
yes me as well

---
In reply to: [400437348](https://github.com/Azure/azure-iot-sdk-python/pull/520#discussion_r400437348) [](ancestors = 400437348)

Reply:
I factored this into its own function because I needed to run the same code when we receive a ConnectedEvent

---
In reply to: [400440435](https://github.com/Azure/azure-iot-sdk-python/pull/520#discussion_r400440435) [](ancestors = 400440435,400437348)